### PR TITLE
[NFS][P0] Add P0 tests for nfs

### PR DESCRIPTION
**Pull request by pranavprakash20**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by pranavprakash20** [comment]
Run logs: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-89DREV/
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-E05JYC/

File: tests/nfs/nfs_verify_pynfs.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    # Step 5: Create a file on Client 1
+    cmd = (
+        f"python3 -m pip install ply;cd {nfs_mount};git clone git://linux-nfs.org/~bfields/pynfs.git;cd pynfs;-- yes | "
+        f"python setup.py build;cd nfs{version};./testserver.py {nfs_server_name}:/testvolume -v --outfile "
+        f"~/pynfs.run --maketree --showomit --rundep all > /tmp/pynfs-hotfix.log"
+    )
+    try:
+        clients[0].exec_command(cmd=cmd, sudo=True)
+    except Exception as e:
```
Comment (msainiRedhat) [inline tests/nfs/nfs_verify_pynfs.py:58]: Need to check the error handling.
Even when testcase failed in logs, it shows "Successfully completed pynfs tests on nfs cluster"

File: tests/nfs/nfs_verify_file_lock.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+    # Wait for the lock to release
+    c1.join()
+
+    # Try again when the lock is released
+    try:
+        clients[1].exec_command(cmd=CMD, sudo=True)
+        log.info(
+            "Expected: Successfully acquired lock from client 2 while client 1 lock is in on"
```
Comment (msainiRedhat) [inline tests/nfs/nfs_verify_file_lock.py:93]: Modify - Expected: Successfully acquired lock from client 2 while client 1 lock is released

Comment (pranavprakash20) [reply tests/nfs/nfs_verify_file_lock.py:93]: Done

File: tests/nfs/nfs_verify_pynfs.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        client.exec_command(cmd=f"mkdir -p {nfs_mount}", sudo=True)
+        cmd = f"mount -t nfs -o vers={version},port={port} {nfs_server_name}:{nfs_export} {nfs_mount}"
+        out, _ = client.exec_command(cmd=cmd, sudo=True)
+        if out:
+            raise OperationFailedError(f"Failed to mount nfs on {client.hostname}")
+
+    # Step 5: Create a file on Client 1
+    cmd = (
+        f"python3 -m pip install ply;cd {nfs_mount};git clone git://linux-nfs.org/~bfields/pynfs.git;cd pynfs;-- yes | "
+        f"python setup.py build;cd nfs{version};./testserver.py {nfs_server_name}:/testvolume -v --outfile "
```
[the remark this answers is not in the document]
Comment (pranavprakash20) [reply tests/nfs/nfs_verify_pynfs.py:53]: Done

File: tests/nfs/nfs_verify_pynfs.py (hunk 1 above)
Comment (pranavprakash20) [reply tests/nfs/nfs_verify_pynfs.py:58]: Noted, check added

File: tests/nfs/nfs_verify_file_lock.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+from time import sleep
+
+from cli.ceph.ceph import Ceph
+from cli.exceptions import OperationFailedError
+from utility.log import Log
+
+log = Log(__name__)
+
+
+CMD = """python3 -c 'from fcntl import flock, LOCK_EX, LOCK_NB, LOCK_UN;from time import sleep;f = open(
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_file_lock.py:11]: Add method for this one in `cli/utils`

File: tests/nfs/nfs_verify_bonnie.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    try:
+        # Step 1: Enable nfs
+        Ceph(clients[0]).mgr.module(action="enable", module="nfs", force=True)
+
+        # Step 2: Create an NFS cluster
+        nfs_server_name = nfs_node.hostname
+        Ceph(clients[0]).nfs.cluster.create(name=nfs_name, nfs_server=nfs_server_name)
+
+        # Step 3: Perform Export on clients
+        for client in [clients[0], clients[1]]:
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_bonnie.py:36]: can be changed to `clients[:1]`

File: tests/nfs/nfs_verify_bonnie.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            Ceph(client).nfs.export.create(
+                fs_name=fs_name, nfs_name=nfs_name, nfs_export=nfs_export, fs=fs
+            )
+
+        # Step 4: Perform nfs mount
+        for client in clients:
+            # Create mount dirs
+            client.exec_command(cmd=f"mkdir -p {nfs_mount}", sudo=True)
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_bonnie.py:44]: These are common steps and can be reused, better to have separate methods in `cli/utils`

File: tests/nfs/nfs_verify_bonnie.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                f"bonnie++ -d {folder_name} -u cephuser",
+            ]
+            ctr += 1
+            for cmd in cmds:
+                client.exec_command(cmd=cmd, sudo=True)
+        log.info("Successfully completed pynfs tests on nfs cluster")
+    except Exception as e:
+        log.error(f"Error : {e}")
+    finally:
+        log.info("Cleaning up")
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_bonnie.py:76]: Is this cleanup ?

File: tests/nfs/nfs_verify_file_lock.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+"/mnt/nfs/sample_file", "w");flock(f.fileno(), LOCK_EX | LOCK_NB);sleep(30);flock(f.fileno(), LOCK_UN)'"""
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_file_lock.py:12]: for sleep method use interval variable

File: tests/nfs/nfs_verify_file_lock.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    nfs_export = "/export"
+    nfs_mount = "/mnt/nfs"
+    fs = "cephfs"
+    for client in [clients[0], clients[1]]:
+        Ceph(client).nfs.export.create(
+            fs_name=fs_name, nfs_name=nfs_name, nfs_export=nfs_export, fs=fs
+        )
+
+    # Step 4: Perform nfs mount
+    for client in [clients[0], clients[1]]:
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_file_lock.py:57]: can be changed to `clients[:1]`

File: tests/nfs/nfs_verify_file_lock.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        # Create mount dirs
+        client.exec_command(cmd=f"mkdir -p {nfs_mount}", sudo=True)
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_file_lock.py:59]: separate methods

File: tests/nfs/nfs_verify_file_lock.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    # Step 5: Create a file on Client 1
+    file_path = "/mnt/nfs/sample_file"
+    clients[0].exec_command(cmd=f"touch {file_path}", sudo=True)
+
+    # Step 6: Perform File Lock from client 1
+    c1 = Thread(target=get_file_lock, args=(clients[0],))
+    c1.start()
+    log.info("Acquired lock from client 1")
+    sleep(5)
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_file_lock.py:73]: Avoid constant sleep. Add proper justification.

File: tests/nfs/nfs_verify_pynfs.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            # Create mount dirs
+            client.exec_command(cmd=f"mkdir -p {nfs_mount}", sudo=True)
+            cmd = f"mount -t nfs -o vers={version},port={port} {nfs_server_name}:{nfs_export} {nfs_mount}"
+            out, _ = client.exec_command(cmd=cmd, sudo=True)
+            if out:
+                raise OperationFailedError(f"Failed to mount nfs on {client.hostname}")
+
+        # Step 5: Create a file on Client 1
+        cmd = (
+            f"python3 -m pip install ply;cd {nfs_mount};git clone git://linux-nfs.org/~bfields/pynfs.git;cd pynfs;-- "
```
Comment (vamahaja) [inline tests/nfs/nfs_verify_pynfs.py:52]: Separate methods

**Review by vamahaja — CHANGES_REQUESTED** [review]
Most of the test case steps are common. Better to have single test file with multiple configs.

File: tests/nfs/nfs_verify_bonnie.py (hunk 5 above)
Comment (pranavprakash20) [reply tests/nfs/nfs_verify_bonnie.py]: Done

File: tests/nfs/nfs_verify_bonnie.py (hunk 7 above)
Comment (pranavprakash20) [reply tests/nfs/nfs_verify_bonnie.py:76]: Yes, cleanup after each test scenario

File: tests/nfs/nfs_verify_file_lock.py (hunk 11 above)
Comment (pranavprakash20) [reply tests/nfs/nfs_verify_file_lock.py]: changed

**Comment by pranavprakash20** [comment]
@vamahaja Changes requested are made. The scope of moving out the common methods to utils can be targeted for a separate PR.

**Review by msainiRedhat — APPROVED** [review]
LGTM

**Review by vamahaja — APPROVED** [review]

**Pull request merged**